cirq-core/cirq/transformers/stratify.py

```diff
@@ -116,43 +113,101 @@ def _stratify_circuit(
     Returns:
         The stratified circuit.
     """
-    num_categories = len(classifiers) + 1
-
-    def map_func(m: 'cirq.Moment', _) -> Sequence['cirq.Moment']:
-        stratified_ops: List[List['cirq.Operation']] = [[] for _ in range(num_categories)]
-        for op in m:
-            if set(op.tags) & set(context.tags_to_ignore):
-                stratified_ops[0].append(op)
-                continue
-            for i, classifier in enumerate(classifiers):
-                if classifier(op):
-                    stratified_ops[i + 1].append(op)
-                    break
-        return [circuits.Moment(op_list) for op_list in stratified_ops]
-
-    stratified_circuit = transformer_primitives.map_moments(circuit, map_func).unfreeze(copy=False)
-    assert len(stratified_circuit) == len(circuit) * num_categories
-
-    # Try to move operations to the left to reduce circuit depth, preserving stratification.
-    for curr_idx, moment in enumerate(stratified_circuit):
-        curr_category = curr_idx % num_categories
-        if curr_category == 0:
-            # Moment containing tagged operations to be ignored.
-            continue
-        batch_removals: List[Tuple[int, 'cirq.Operation']] = []
-        batch_inserts: List[Tuple[int, 'cirq.Operation']] = []
+    num_classes = len(classifiers) + 1  # include one "extra" category for ignored operations
+    new_moments: List[List['cirq.Operation']] = []
+
+    # Keep track of the the latest time index addressing a qubit.
+    qubit_time_index = {qubit: -1 for qubit in circuit.all_qubits()}
+
+    # Keep track of the latest time index addressing a measurement/control key.
+    measurement_time_index: Dict['cirq.MeasurementKey', int] = {}
+    control_time_index: Dict['cirq.MeasurementKey', int] = {}
+
+    # The minimum time index for operations with a tag in context.tags_to_ignore.
+    min_time_index_for_ignored_op = 0
```

Comment:
Consider renaming `min_time_index_for_ignored_op` to `last_ignored_ops_moment_index` because we are essentially keeping track of the latest moment that contains only ignored ops so that we can insert the next ignores ops batch after that time. 

Also, compute `min_time_index_for_ignored_op` as just the `max(min_time_index_for_ignored_op, time_index)` in the inner loop and get rid of the check `min_time_index_for_op = max(min_time_index_for_op, min_time_index_for_ignored_op)`. Once you have `min_time_index_for_ignored_op` in the inner loop, you can take the max with `last_ignored_ops_moment_index` once outside before assigning `op_time_indices` to `ignored_ops` and then update the `last_ignored_ops_moment_index` with the current index. 


In general, I think the current logic to take care of ignored ops is a bit convoluted and it's hard to reason about it's correctness. My proposed changes should make it easier for a reader to quickly understand how we are handling the `ignored_ops`.

Reply:
Thanks for these suggestions for how to clean up `ignored_op` handing.  I'm not sure I implemented them exactly as requested, but the new version is certainly much cleaner.  What do you think?